File: migrations/r0/m3/s1-nextjs/code/migrated-project/src/lib/repositories/donor.repository.ts
import { execute, query, queryOne } from "@/lib/db/pool";
import type {
  Donor,
  DonorInput,
  DonorPendingReceipt,
  DonorSelect,
  Item,
} from "@/lib/types";

export async function findAllDonors(): Promise<Donor[]> {
  return query<Donor>(
    "SELECT * FROM Donor ORDER BY BusinessName, ContactName",
  );
}

export async function findDonorById(donorId: number): Promise<Donor | null> {
  return queryOne<Donor>("SELECT * FROM Donor WHERE DonorID = ?", [donorId]);
}

export async function findDonorsForSelect(): Promise<DonorSelect[]> {
  return query<DonorSelect>(
    "SELECT DonorID, BusinessName, ContactName FROM Donor ORDER BY BusinessName, ContactName",
  );
}

export async function findDonorsWithoutReceipt(): Promise<DonorPendingReceipt[]> {
  return query<DonorPendingReceipt>(
    `SELECT d.*, COUNT(i.ItemID) AS TotalItems, SUM(i.RetailValue) AS TotalValue
     FROM Donor d
     INNER JOIN Item i ON d.DonorID = i.DonorID
     WHERE d.TaxReceipt = 0
     GROUP BY d.DonorID
     HAVING TotalItems > 0
     ORDER BY d.ContactName, d.BusinessName`,
  );
}

export async function findItemsByDonorId(donorId: number): Promise<Item[]> {
  return query<Item>(
    "SELECT * FROM Item WHERE DonorID = ? ORDER BY Description",
    [donorId],
  );
}

export async function donorHasItems(donorId: number): Promise<boolean> {
  const result = await queryOne<{ count: number }>(
    "SELECT COUNT(*) AS count FROM Item WHERE DonorID = ?",
    [donorId],
  );
  return (result?.count ?? 0) > 0;
}

export async function createDonor(values: DonorInput): Promise<void> {
  await execute(
    `INSERT INTO Donor (
       BusinessName, ContactName, ContactEmail, ContactTitle,
       Address, City, State, ZipCode, TaxReceipt
     ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
    [
      values.businessName ?? null,
      values.contactName,
      values.contactEmail ?? "",
      values.contactTitle ?? null,
      values.address ?? "",
      values.city ?? "",
      values.state ?? "",
      values.zipCode ?? "",
      values.taxReceipt ? 1 : 0,
    ],
  );
}

export async function updateDonor(
  donorId: number,
  values: DonorInput,
): Promise<void> {
  await execute(
    `UPDATE Donor SET
       BusinessName = ?, ContactName = ?, ContactEmail = ?, ContactTitle = ?,
       Address = ?, City = ?, State = ?, ZipCode = ?, TaxReceipt = ?
     WHERE DonorID = ?`,
    [
      values.businessName ?? null,
      values.contactName,
      values.contactEmail ?? "",
      values.contactTitle ?? null,
      values.address ?? "",
      values.city ?? "",
      values.state ?? "",
      values.zipCode ?? "",
      values.taxReceipt ? 1 : 0,
      donorId,
    ],
  );
}

export async function deleteDonor(donorId: number): Promise<void> {
  await execute("DELETE FROM Donor WHERE DonorID = ?", [donorId]);
}

export async function markReceiptSent(donorId: number): Promise<void> {
  await execute("UPDATE Donor SET TaxReceipt = 1 WHERE DonorID = ?", [donorId]);
}

File: migrations/r0/m3/s2-blazor/code/migrated-project/SilentAuction/Data/Repositories/DonorRepository.cs
using Microsoft.EntityFrameworkCore;
using SilentAuction.Data.Entities;
using SilentAuction.Models;

namespace SilentAuction.Data.Repositories;

public class DonorRepository(SilentAuctionDbContext db) : IDonorRepository
{
    public async Task<List<Donor>> GetAllAsync() =>
        await db.Donors
            .OrderBy(d => d.BusinessName)
            .ThenBy(d => d.ContactName)
            .ToListAsync();

    public async Task<Donor?> GetByIdAsync(int donorId) =>
        await db.Donors.FindAsync(donorId);

    public async Task<List<DonorSelectDto>> GetForSelectAsync() =>
        await db.Donors
            .OrderBy(d => d.BusinessName)
            .ThenBy(d => d.ContactName)
            .Select(d => new DonorSelectDto
            {
                DonorID = d.DonorID,
                BusinessName = d.BusinessName,
                ContactName = d.ContactName
            })
            .ToListAsync();

    public async Task<List<DonorPendingReceiptDto>> GetWithoutReceiptAsync() =>
        await db.Donors
            .Where(d => !d.TaxReceipt)
            .SelectMany(d => d.Items, (d, i) => new { Donor = d, Item = i })
            .GroupBy(x => x.Donor.DonorID)
            .Select(g => new DonorPendingReceiptDto
            {
                DonorID = g.First().Donor.DonorID,
                BusinessName = g.First().Donor.BusinessName,
                ContactName = g.First().Donor.ContactName,
                ContactEmail = g.First().Donor.ContactEmail,
                ContactTitle = g.First().Donor.ContactTitle,
                Address = g.First().Donor.Address,
                City = g.First().Donor.City,
                State = g.First().Donor.State,
                ZipCode = g.First().Donor.ZipCode,
                TaxReceipt = g.First().Donor.TaxReceipt,
                TotalItems = g.Count(),
                TotalValue = g.Sum(x => x.Item.RetailValue)
            })
            .OrderBy(d => d.ContactName)
            .ThenBy(d => d.BusinessName)
            .ToListAsync();

    public async Task<List<Item>> GetItemsByDonorIdAsync(int donorId) =>
        await db.Items
            .Where(i => i.DonorID == donorId)
            .OrderBy(i => i.Description)
            .ToListAsync();

    public async Task<bool> DonorHasItemsAsync(int donorId) =>
        await db.Items.AnyAsync(i => i.DonorID == donorId);

    public async Task<bool> AddAsync(Donor donor)
    {
        db.Donors.Add(donor);
        return await SaveAsync();
    }

    public async Task<bool> UpdateAsync(Donor donor)
    {
        db.Donors.Update(donor);
        return await SaveAsync();
    }

    public async Task<bool> DeleteAsync(int donorId)
    {
        var donor = await db.Donors.FindAsync(donorId);
        if (donor is null) return false;
        db.Donors.Remove(donor);
        return await SaveAsync();
    }

    public async Task<bool> MarkReceiptSentAsync(int donorId)
    {
        var donor = await db.Donors.FindAsync(donorId);
        if (donor is null) return false;
        donor.TaxReceipt = true;
        return await SaveAsync();
    }

    private async Task<bool> SaveAsync()
    {
        try
        {
            await db.SaveChangesAsync();
            return true;
        }
        catch
        {
            return false;
        }
    }
}
